### PR TITLE
Check if last tested and minimum NVDA versions exist in available API versions

### LINKED ISSUE #16: Check that last tested version exists in NVDA
In addonFiles repo,manual reviews need to ensure that last tested version exist in NVDA:
- When a beta is published, the corresponding last tested version is considered valid, but versions cannot be greater than it, for example, a version like 9999.1.0 would be invalid, or even 2024.1.0. This is a way to skipping the compatibility procedure set in NVDA.
- Seems that this is not checked, so that we may release add-ons with a big version to avoid re-tests and updates when NVDA breaks compatibility.

A maxApiVersion may be provided in this repo, or for better maintainment, addon-datastore-transform maybe checked out in GitHub Actions to see the last version existing in NVDAApiVersions.

cc: @seanbudd

### PR DESCRIPTION
**Pull request by nvdaes**
Closes #16

### CODE REVIEW

**Comment by nvdaes** [comment]
@seanbudd If this is useful for you, feel free to review it.
Unfortunately there's a test failing and I don't know how to fix it. But manual tests with add-ons are good for me.

**Comment by nvdaes** [comment]
This fixes issue #16 

- nvdaAPIVersions.json available in addons-datastore-transform repo is downloaded in the temp dir each time a metadata file is validated. This avoids to store this file also in this repo, and ensures it's updated. Also avoids an extra argument, i.e., it's not needed to specify where the API version file is located in the command line.
- Tests are added for the new functions, though unfortunately something in E2E tests is failing now.
- The readme has been updated to inform that min and last test versions are checked against API versions.

File: README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ The Action aims to validate the metadata of add-ons submitted to
   - The submission addonName matches the manifest summary field
   - The submission description matches the manifest description field
   - The homepage URL matches the manifest URL field
-  - The addon versions match.
+  - The addon versions match, and last tested & minimum required versionsexist in NVDA API versions.
```
Comment (seanbudd) [inline README.md]: add space

```suggestion
  - The addon versions match, and last tested & minimum required versions exist in NVDA API versions.
```

File: lint.txt [hunk 2]
```diff
@@ -0,0 +1,74 @@
+Ensuring NVDA Python virtual environment
```
Comment (seanbudd) [inline lint.txt]: this file should be removed

File: nvdaAPIVersions.json [hunk 3]
```diff
@@ -0,0 +1,327 @@
+[
```
Comment (seanbudd) [inline nvdaAPIVersions.json]: i don't think we should host a copy of this file here.
Instead, the dependency should be passed in via command line e.g. `runvalidate <pathToAddonMetadataFile.json> <pathTonvdaAPIVersions.json>`

File: _validate/validate.py [hunk 4]
```diff
@@ -23,9 +23,9 @@
 
 
 JSON_SCHEMA = os.path.join(os.path.dirname(__file__), "addonVersion_schema.json")
-JsonObjT = typing.Dict[str, typing.Any]
-
+VER_URL = "https://raw.githubusercontent.com/nvaccess/addon-datastore-transform/main/nvdaAPIVersions.json"
```
Comment (seanbudd) [inline _validate/validate.py]: i don't think we should download a copy like this.
Instead, the dependency should be passed in via command line e.g. `runvalidate <pathToAddonMetadataFile.json> <pathTonvdaAPIVersions.json>`

**Review by seanbudd — COMMENTED** [review]
Hi @nvdaes,

Thanks for working on this.
I've given some initial suggestions before performing an in-depth review

**Comment by nvdaes** [comment]
Ok, i"ll work on this tomorrow.yEnviado desde mi iPhoneEl 2 mar 2023, a las 20:16, Sean Budd ***@***.***> escribió:﻿
@seanbudd commented on this pull request.

Hi @nvdaes,
Thanks for working on this.
I've given some initial suggestions before performing an in-depth review

In README.md:
> @@ -17,7 +17,7 @@ The Action aims to validate the metadata of add-ons submitted to
   - The submission addonName matches the manifest summary field
   - The submission description matches the manifest description field
   - The homepage URL matches the manifest URL field
-  - The addon versions match.
+  - The addon versions match, and last tested & minimum required versionsexist in NVDA API versions.

add space
⬇️ Suggested change
-  - The addon versions match, and last tested & minimum required versionsexist in NVDA API versions.
+  - The addon versions match, and last tested & minimum required versions exist in NVDA API versions.





In lint.txt:
> @@ -0,0 +1,74 @@
+Ensuring NVDA Python virtual environment

this file should be removed



In nvdaAPIVersions.json:
> @@ -0,0 +1,327 @@
+[

i don't think we should host a copy of this file here.
Instead, the dependency should be passed in via command line e.g. runvalidate <pathToAddonMetadataFile.json> <pathTonvdaAPIVersions.json>



In _validate/validate.py:
> @@ -23,9 +23,9 @@
 
 
 JSON_SCHEMA = os.path.join(os.path.dirname(__file__), "addonVersion_schema.json")
-JsonObjT = typing.Dict[str, typing.Any]
-
+VER_URL = "https://raw.githubusercontent.com/nvaccess/addon-datastore-transform/main/nvdaAPIVersions.json"

i don't think we should download a copy like this.
Instead, the dependency should be passed in via command line e.g. runvalidate <pathToAddonMetadataFile.json> <pathTonvdaAPIVersions.json>

**Comment by nvdaes** [comment]
Finally I'have been able to work today on this and, after addressing your comments, it seems to work. I'll wait for further review.

File: _tests/testData/nvdaAPIVersions.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
+[
```
Comment (seanbudd) [inline _tests/testData/nvdaAPIVersions.json:1]: I think keeping this file as short as possible like it was earlier is preferred. The simpler the test case the better, while still covering all the edge cases (e.g. too low, too high).

File: _validate/validate.py [hunk 6]
```diff
@@ -39,6 +38,14 @@ def getAddonMetadata(filename: str) -> JsonObjT:
 	return data
 
 
+def getExistingVersions(verFilename: str) -> typing.List[str]:
+	"""Loads API versions file and returns list of versions formatted as strings.
+	"""
+	with open(verFilename) as f:
+		data: JsonObjT = json.load(f)
+	return (_formatVersionString(version["apiVer"].values()) for version in data)
```
Comment (seanbudd) [inline _validate/validate.py]: this does not return a list, you may want to wrap it as a tuple or list directly

File: _validate/validate.py [hunk 7]
```diff
@@ -223,22 +230,39 @@ def checkManifestVersionMatchesVersionName(
...
 
 
+def checkLastTestedVersionExist(submission: JsonObjT, verFilename: str) -> ValidationErrorGenerator:
+	lastTestedVersion: JsonObjT = submission['lastTestedVersion']
+	formattedLastTestedVersion: str = _formatVersionString(lastTestedVersion.values())
+	if formattedLastTestedVersion not in getExistingVersions(verFilename):
+		yield f"Last tested version error: {formattedLastTestedVersion} doesn't exist"
+
+
+def checkMinRequiredVersionExist(submission: JsonObjT, verFilename: str) -> ValidationErrorGenerator:
```
Comment (seanbudd) [inline _validate/validate.py:241]: I think this check should be the same as `checkLastTestedVersionExist`, it should check if the version exists and nothing else.
I'm not sure I understand the current logic.

File: _validate/validate.py [hunk 8]
```diff
@@ -223,22 +230,39 @@ def checkManifestVersionMatchesVersionName(
...
-	"""
+	"""Check submitted json file name matches the *.nvda-addon manifest name field."""
+
 	yield from checkSubmissionFilenameMatchesVersionNumber(
 		submissionFilePath,
 		submission
 	)
+
 	yield from checkManifestVersionMatchesVersionName(manifest, submission)
 	yield from checkParsedVersionNameMatchesVersionNumber(submission)
```
Comment (seanbudd) [inline _validate/validate.py]: these changes should probably be reverted

**Review by seanbudd — COMMENTED** [review]
Hi @nvdaes , this is looking almost ready

**Comment by seanbudd** [comment]
Could you also update the readme to discuss the new usage convention?

**Comment by nvdaes** [comment]
Hi @seanbudd Thanks for your review

- I'll update the readme.
- For the getExistingVersions function, I'lreturn directly a listfor simplicity.
- In checkLastTestedVersionExist function, I think we have to check if the tested version is old and then consider its formatted version as "0.0.0", like done in checkLastMinRequiredVersion. Should we add a function to convert, for example, version 2018.1 to API version 0.0.0? I've trying removing this conversion, as I have understood from your suggestion, and tests fail.
- For the nvdaAPIVersions.json file, I'll remove some entries. But we have to maintain some of them like 2919.3.0 and 2020.4 since they seem to be used when running end to end tests with the fake add-on stored in testData. Please let me know if you have further comments or concerns about this.

**Comment by seanbudd** [comment]
2018.1 is an invalid version. Instead authors should submit 0.0.0.

It might be worth updating those tests to make the number of versions more
minimal.
But whatever is easiest, this is quite minor.

**Comment by nvdaes** [comment]
I know that there are more changes to do and I may finish this in a few hours.

**Comment by nvdaes** [comment]
Hi @seanbudd: I think that all is addressed. Let me know if I should do anything else.
Thanks.

**Comment by CyrilleB79** [comment]
Sorry in case I have not understand things correctly.

But to me, min version should be any version of NVDA, not only API-breaking version. See https://github.com/nvaccess/addon-datastore-validation/issues/16#issuecomment-1454911881 for the explanation. And see Character Information compatible with NVDA 2022.3.3 onwards for an example.

**Comment by nvdaes** [comment]
Cyrille, note that api versions aren"t api breaking changes versions. For enample, 2020.-4 is a valid api version and this doesn"t break changes. Valid api versions are: 0. 0. 0, for 2019 dot 2 and lower. 2019.-3, 2020.-1, 2020.-2, etc. Please see addn-datastore-transform nvdaapiversions.json. There you"ll se an array of dictionaries, and each entry contains a description, an apIVer and a backCompactTo key. Perhaps i am missing something of your explanation but I thing that it"s not needed to include in nvdaapiversions file all api versions like 2016.-1, 2017.-3 and so on, and it"s difficult to say what versions were released, of NVDA, if they aren"t listed in a file. In this case i don"t agree with you.Enviado desde mi iPhoneEl 4 mar 2023, a las 23:51, Cyrille Bougot ***@***.***> escribió:﻿
Sorry in case I have not understand things correctly.
But to me, min version should be any version of NVDA, not only API-breaking version. See #16 (comment) for the explanation. And see Character Information compatible with NVDA 2022.3.3 onwards for an example.

File: README.md [hunk 9]
```diff
@@ -17,7 +17,7 @@ The Action aims to validate the metadata of add-ons submitted to
   - The submission addonName matches the manifest summary field
   - The submission description matches the manifest description field
   - The homepage URL matches the manifest URL field
-  - The addon versions match.
+  - The addon versions match, and last tested & minimum required versions exist in NVDA API versions.
```
Comment (seanbudd) [inline README.md]: can this be on a new line. 
```suggestion
  - The addon versions match
  - The last tested & minimum required versions are valid NVDA API versions.
```

File: _validate/validate.py [hunk 10]
```diff
@@ -23,9 +23,8 @@
 
 
 JSON_SCHEMA = os.path.join(os.path.dirname(__file__), "addonVersion_schema.json")
-JsonObjT = typing.Dict[str, typing.Any]
-
 
+JsonObjT = typing.Dict[str, typing.Any]
```
Comment (seanbudd) [inline _validate/validate.py]: can this diff be removed?

File: _validate/validate.py [hunk 11]
```diff
@@ -291,12 +315,17 @@ def main():
 		dest="file",
 		help="The json (.json) file containing add-on metadata."
 	)
+	parser.add_argument(
+		dest="versions",
```
Comment (seanbudd) [inline _validate/validate.py]: APIVersions I think would be a better name for this

```suggestion
		dest="APIVersions",
```

File: _validate/validate.py [hunk 12]
```diff
@@ -291,12 +315,17 @@ def main():
...
+		help="The json (.json) file containing API versions."
```
Comment (seanbudd) [inline _validate/validate.py]:
```suggestion
		help="The JSON file containing valid NVDA API versions."
```

**Review by seanbudd — COMMENTED** [review]
Thanks Noelia, this looks almost ready. Just some trivial changes remaining.

**Comment by nvdaes** [comment]
@seanbudd I think I'll addressed all your comments. Let me know if I should do anything else.

File: _validate/validate.py [hunk 13]
```diff
@@ -23,8 +23,8 @@
 
 
 JSON_SCHEMA = os.path.join(os.path.dirname(__file__), "addonVersion_schema.json")
-JsonObjT = typing.Dict[str, typing.Any]
 
+JsonObjT = typing.Dict[str, typing.Any]
```
Comment (seanbudd) [inline _validate/validate.py]: there's still a diff here, not sure why

**Comment by nvdaes** [comment]
Can you explain what's the diff, for example, an extra line? I'm not sure.

2023-03-06 7:01 GMT+01:00, Sean Budd ***@***.***>:

**Comment by nvdaes** [comment]
@seanbudd I've searched better the diff and now it should be fixed.

File: _validate/validate.py [hunk 14]
```diff
@@ -291,12 +316,17 @@ def main():
...
 		help="The json (.json) file containing add-on metadata."
 	)
+	parser.add_argument(
+		dest="APIVersions",
+		help="The JSON file containing valid NVDA API versions."
+	)
 
 	args = parser.parse_args()
 	filename = args.file
+	verFilename = args.versions
```
Comment (seanbudd) [inline _validate/validate.py]: doesn't this need to be changed as well?

```suggestion
	verFilename = args.APIVersions
```

**Comment by nvdaes** [comment]
Hi @seanbudd I have addressed your last suggestion.

**Review by seanbudd — APPROVED** [review]
